tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala

```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+          val idempotent = isIdempotent(inst.op)
+
+          val instHash =
+            if (idempotent)
+              hash(inst.op)
+            else
+              inst.hashCode // hash the assigned variable as well, so a = op(b) and c = op(b) don't have the same hash
+
+          variableVN += (inst.name -> instHash)
+          localDefs += (inst.name  -> inst)
+
+          if (idempotent) {
+            val hashEqualInstrs = instructions.getOrElse(instHash, Nil)
+            instructions += (instHash -> (inst :: hashEqualInstrs))
+
+            val equalInstrs =
+              hashEqualInstrs.filter(otherInst =>
+                deepEquals.eqInst(inst, otherInst))
+            val redundantInstrs = equalInstrs.filter(eqInst =>
+              blockDominatedByDef(block, eqInst.name)) // only redundant if the current block is dominated by the block in which the equal instruction occurs
+
+            val newInstOpt = redundantInstrs.headOption.map(
+              redInst =>
+                Inst.Let(inst.name,
+                         Op.Copy(Val.Local(redInst.name, redInst.op.resty))))
+            newInstOpt.getOrElse(inst)
+          } else {
+            inst
+          }
+        }
+
+        case otherInst @ _ =>
+          otherInst
+      }
+
+      block.label +: newBlockInsts
+    }
+
+    newInsts.flatten
+  }
+
+}
+
+object GlobalValueNumbering extends PassCompanion {
+  def apply(ctx: Ctx) = new GlobalValueNumbering()(ctx.top)
+
+  /* The definition of idempotent for Call might be wrong.
+   * Example:
+   * val a = new AnyRef { var b: Int = 0; override def toString = b.toString }
+   * val c = a.toString
+   * a.b = 1
+   * val d = a.toString
+   * --> toString is pure, but the values of c and d are different !
+   * */
+  def isIdempotent(op: Op)(implicit top: Top): Boolean = {
+    import Op._
+    op match {
+      // Always idempotent:
+      case (_: Pure | _: Method | _: As | _: Is | _: Copy | _: Sizeof |
+          _: Module | _: Field) =>
+        true
+
+      // Never idempotent:
+      case (_: Load | _: Store | _: Stackalloc | _: Classalloc) => false
+
+      // Special cases:
+      case Call(_, Val.Global(Ref(node), _), _) =>
+        node.attrs.isPure
+
+      case Call(_, _, _) => false
+
+      case _: Closure => ???
+    }
+  }
+
+  class DeepEquals(localDefs: Local => Inst)(implicit top: Top) {
+
+    def eqInst(instA: Inst.Let, instB: Inst.Let): Boolean = {
+      (instA.name == instB.name) || eqOp(instA.op, instB.op)
+    }
+
+    def eqOp(opA: Op, opB: Op): Boolean = {
+      import Op._
+      if (!(isIdempotent(opA) && isIdempotent(opB)))
+        false
+      else {
+        (opA, opB) match {
+          // Here we know that the function called is idempotent
+          case (Call(tyA, ptrA, argsA), Call(tyB, ptrB, argsB)) =>
+            eqType(tyA, tyB) && eqVal(ptrA, ptrB) && eqVals(argsA, argsB)
+
+          case (Load(_, _), Load(_, _)) =>
+            false // not idempotent
```

Comment:
It seems that this and all the other not-idempotent cases are never going to happen due to the check above, therefore we should drop this code altogether.